mesh overhaul

tomvanmele opened the pull request:
<!-- Thank you for your pull request!  -->
<!-- Please start by describing your change in a few sentences. -->
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [ ] New feature in a **backwards-compatible** manner.
- [x] Breaking change: bug fix or new feature that involve incompatible API changes.

### Checklist

1. [x] Add the change to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading: `Added`, `Changed`, `Removed`.
1. [x] Run all tests on your computer (i.e. `invoke test`).
1. [ ] If you add new functions/classes, check that:
   1. [ ] Are available on a second-level import, e.g. `compas.datastructures.Mesh`.
   1. [ ] Add unit tests (especially important for algorithm implementations).

tomvanmele committed:
mesh overhaul

tomvanmele committed:
remove test file

tomvanmele commented:
* mesh attribute getter and setter functions have been merged into one.
* only modified attributes are actually stored
* iterators return a mutable view of a combination of default and modified attributes
* standard mesh algorithms are added to a public n-gon mesh class
* the base mesh implementation and related base functionality have been moved to a core submodule

tomvanmele committed:
updating docs

gonzalocasas commented on src/compas/datastructures/mesh/core/basemesh.py:
```diff
@@ -42,30 +43,127 @@
+        self.attr = attr
+
+    def __getitem__(self, key):
+        if key in self.attr:
```
How about changing this part to assume it is there, and move the fallback behavior to an `except`? something like:

```python
try:
    return self.attr[key]
except KeyError:
    return self.defaults[key]
```

That way, you don't incur any extra cost for the common case in which the attribute is defined.

gonzalocasas commented on src/compas/datastructures/mesh/core/basemesh.py:
```diff
@@ -42,30 +43,127 @@
-================================================================================
-Mesh summary
-================================================================================
+class AttributeView(object):
```
Although internal, these classes (`*AttributeView`) would need a few lines of doc to clarify intent/purpose.

gonzalocasas commented on src/compas/datastructures/mesh/core/basemesh.py:
```diff
@@ -637,40 +771,49 @@ def from_vertices_and_faces(cls, vertices, faces):
-
         return mesh
 
+    def to_vertices_and_faces(self):
```
I think this method is core enough that it would really require a few unit tests on it (if not already present).

gonzalocasas commented on src/compas/datastructures/mesh/core/basemesh.py:
```diff
@@ -1287,59 +1159,635 @@ def delete_face(self, fkey):
+        
+        Returns
+        -------
+        None
```
Is this really how to document different return values? I would have assumed only inline comment for this, and this would be assumed to be a 3-tuple return.

gonzalocasas reviewed (commented):
Review still in progress but I'll push out a few comments...

tomvanmele committed:
Merge branch 'master' into mesh-overhaul

# Conflicts:
#	CHANGELOG.md
#	src/compas/geometry/queries.py
#	src/compas/numerical/dr/dr_numpy.py

tomvanmele committed:
try-except for vertex attributes
better specification of return values
added keyerror to attribute functions

gonzalocasas commented:
LGTM!

The core `basemesh.py` is really too big to review, but in general, I think it's good to go.

The only thing I would like to be included if possible before merge would be a method to unset an attribute. I think this is now not possible because the `vertices_attribute` method relies on value being something in order to act as a setter (as discussed last week).

tomvanmele committed:
Merge branch 'master' into mesh-overhaul

tomvanmele committed:
clean up of public mesh
added unset methods for vertex/face/edge attributes
replaced has_vertex/edge/face by is_vertex/edge/face
